contracts/ShortNameClaims.sol

```diff
@@ -82,80 +106,193 @@ contract ShortNameClaims is Ownable {
+    function submitPrefixClaim(bytes memory name, address claimant, string memory email) public payable {
         bytes memory firstLabel = bytes(getLabel(name, 0));
         require(firstLabel.equals(firstLabel.length - 3, bytes("eth")));
-        handleClaim(string(firstLabel.substring(0, firstLabel.length - 3)), name, claimant);
+        handleClaim(string(firstLabel.substring(0, firstLabel.length - 3)), name, claimant, email);
+    }
+
+    /**
+     * @dev Closes the claim submission period.
+     *      Callable only by the owner.
+     */
+    function closeClaims() external onlyOwner inPhase(Phase.OPEN) {
+        phase = Phase.REVIEW;
+    }
+
+    /**
+     * @dev Ratifies the current set of claims.
+     *      Ratification freezes the claims and their resolutions, and permits
+     *      them to be acted on.
+     */
+    function ratifyClaims() external inPhase(Phase.REVIEW) {
+        // Only the ratifier can ratify.
+        require(msg.sender == ratifier);
+        // Can't ratify until all claims have a resolution.
+        require(pendingClaims == 0);
+        phase = Phase.FINAL;
     }
 
-    function approveClaim(bytes32 claimId) onlyOwner public {
+    /**
+     * @dev Cleans up the contract, after all claims are resolved.
+     *      Callable only by the owner, and only in final state.
+     */
+    function destroy() external onlyOwner inPhase(Phase.FINAL) {
+        require(unresolvedClaims == 0);
+        selfdestruct(toPayable(ratifier));
+    }
+
+    /**
+     * @dev Sets the status of a claim to either APPROVED or DECLINED.
+     *      Callable only during the review phase, and only by the owner or
+     *      ratifier.
+     * @param claimId The claim to set the status of.
+     * @param status The status to set - must be APPROVED or DECLINED.
+     */
+    function setClaimStatus(bytes32 claimId, Status status) public inPhase(Phase.REVIEW) {
+        // Only callable by owner or ratifier
+        require(msg.sender == owner() || msg.sender == ratifier);
+
+        // Can't set claim back to pending or to withdrawn
+        require(status == Status.APPROVED || status == Status.DECLINED);
+
         Claim memory claim = claims[claimId];
         require(claim.paid > 0, "Claim not found");
 
-        claimCount--;
-        delete claims[claimId];
-        emit ClaimApproved(claimId);
+        if(claim.status == Status.PENDING) {
+          // Claim went from pending -> approved/declined; update counters
+          pendingClaims--;
+          unresolvedClaims++;
+        } else if(claim.status == Status.APPROVED) {
+          // Claim was previously approved; remove from approved map
+          approvedNames[claim.labelHash] = false;
+        }
+
+        // Claim was just approved; check the name was not already used, and add
+        // to approved map
+        if(status == Status.APPROVED) {
+          require(!approvedNames[claim.labelHash]);
+          approvedNames[claim.labelHash] = true;
+        }
```

Comment:
It would be a lot easier to understand if you have the following code to handle the declined case rather than the part I commented at line 206

```
else if (status == Status.DECLINED){
   approvedNames[claim.labelHash] = false;
}

Reply:
If we do that, though, Solidity executes a no-op store of 0, costing us 5k extra gas every time a claim is declined.